### PR TITLE
Deprecate public internals and fix warning suppression 

### PR DESCRIPTION
**Pull request by adkian-sifive**
### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [ ] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [ ] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

  - code cleanup   

#### API Impact

Public APIs in `chisel3.internal` that never should have been public are being deprecated.

#### Backend Code Generation Impact

N/A
#### Desired Merge Strategy

 - Squash: The PR will be squashed and merged (choose this if you have no preference.

#### Release Notes

Internal APIs in package `chisel3.internal` which never should have been public are deprecated.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: core/src/main/scala/chisel3/package.scala [hunk 1]
```diff
@@ -361,4 +361,6 @@ package object chisel3 {
 
   final val deprecatedMFCMessage =
     "this feature will not be supported as part of the migration to the MLIR-based FIRRTL Compiler (MFC). For more information about this migration, please see the Chisel ROADMAP.md."
+
+  final val deprecatedPublicAPIMsg = "APIs in chisel3.internal will no longer be public as of 3.6"
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/package.scala:365]:
```suggestion
  final val deprecatedPublicAPIMsg = "APIs in chisel3.internal are not intended to be public"
```
They'll just be deprecated in 3.6, not private until 5 so I think we just warn and don't mention a version for when they'll be removed.

File: core/src/main/scala/chisel3/internal/Binding.scala [hunk 2]
```diff
@@ -103,42 +112,51 @@ sealed trait ConditionalDeclarable extends TopBinding {
...
 }
 // A DontCare element has a specific Binding, somewhat like a literal.
 // It is a source (RHS). It may only be connected/applied to sinks.
+@deprecated(deprecatedPublicAPIMsg, "Chisel 3.6")
 case class DontCareBinding() extends UnconstrainedBinding
 
 // Views currently only support 1:1 Element-level mappings
-private[chisel3] case class ViewBinding(target: Element) extends UnconstrainedBinding
+@deprecated(deprecatedPublicAPIMsg, "Chisel 3.6")
+case class ViewBinding(target: Element) extends UnconstrainedBinding
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Binding.scala:159]:
```suggestion
private[chisel3] case class ViewBinding(target: Element) extends UnconstrainedBinding
```
This one is already private, why are you making it public?

File: core/src/main/scala/chisel3/internal/Binding.scala [hunk 3]
```diff
@@ -147,20 +165,25 @@ private[chisel3] case class ViewBinding(target: Element) extends UnconstrainedBi
   * @note The types of key and value need not match for the top Data in a total view of type
   *       Aggregate
   */
-private[chisel3] case class AggregateViewBinding(childMap: Map[Data, Data]) extends UnconstrainedBinding {
+@deprecated(deprecatedPublicAPIMsg, "Chisel 3.6")
+case class AggregateViewBinding(childMap: Map[Data, Data]) extends UnconstrainedBinding {
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Binding.scala:169]:
```suggestion
private[chisel3] case class AggregateViewBinding(childMap: Map[Data, Data]) extends UnconstrainedBinding {
```

File: core/src/main/scala/chisel3/internal/Binding.scala [hunk 4]
```diff
@@ -28,7 +33,11 @@ object requireIsHardware {
...
 /** Requires that a node is a chisel type (not hardware, "unbound")
   */
+@deprecated(deprecatedPublicAPIMsg, "Chisel 3.6")
 object requireIsChiselType {
+  def apply(node: Data, msg: String = ""): Unit = _requireIsChiselType.apply(node, msg)
+}
+private[chisel3] object _requireIsChiselType {
   def apply(node: Data, msg: String = ""): Unit = if (node.isSynthesizable) {
     val prefix = if (msg.nonEmpty) s"$msg " else ""
     throw ExpectedChiselTypeException(s"$prefix'$node' must be a Chisel type, not hardware")
```
Comment (jackkoenig) [inline core/src/main/scala/chisel3/internal/Binding.scala:43]: This, and `requireIsHardware` above should just be moved to chisel3.experimental (where there is currently a val alias). Then, instead of objects that replicate all of the arguments, you should just put a deprecated val alias in package object chisel3.internal.

File: build.sbt [hunk 5]
```diff
@@ -36,6 +36,8 @@ lazy val commonSettings = Seq(
   }
 )
 
+lazy val warningSuppression = Seq(scalacOptions += "-Wconf:msg=APIs in chisel3.internal:s,msg=Importing from firrtl:s,msg=migration to the MLIR:s")
```
Comment (jackkoenig) [inline build.sbt:39]:
```suggestion
lazy val warningSuppression = Seq(
  scalacOptions += "-Wconf:" + Seq(
    "msg=APIs in chisel3.internal:s",
    "msg=Importing from firrtl:s",
    "msg=migration to the MLIR:s"
  ).mkString(",")
)
```
I think we should write this in a way that highlights each rule being applied.

**Comment by adkian-sifive** [comment]
Note: SourceInfo will be moved to package experimental in a follow up PR. This is because moving around packages causes the chiseltest integration test CI checks to fail, and we don't currently have the ability to skip this check in CI. The follow up PR will happen after the ability to skip chiseltest is added in a separate PR.

File: build.sbt [hunk 6]
```diff
@@ -17,9 +17,9 @@ lazy val commonSettings = Seq(
   organization := "edu.berkeley.cs",
   version := "3.6-SNAPSHOT",
   autoAPIMappings := true,
+  scalacOptions := Seq("-deprecation", "-feature"),
   scalaVersion := "2.12.17",
   crossScalaVersions := Seq("2.13.10", "2.12.17"),
-  scalacOptions := Seq("-deprecation", "-feature"),
```
Comment (jackkoenig) [inline build.sbt:22]: Can you remove this unnecessary diff?

**Review by jackkoenig — APPROVED** [review]
1 minor request but LGTM otherwise!

File: build.sbt (hunk 6 above)
Comment (jackkoenig) [reply build.sbt]: I pushed a commit that fixes it.

**Comment by jackkoenig** [comment]
Also thank you very very much for cleaning up the warnings, it reduced the total number of warnings for a Chisel3 compile from 188 to 59.
